Issue/4717

axel22 opened the pull request:
Fixes SI-4717.
A lazy val declared inside an anonymous class inside a specialized
context no longer crashes Duplicators.

Previously, a duplicated lazy val was assigned to the wrong owner
in Duplicators:

```
class Foo[@specialized(Boolean) A] {
  def x[B >: A]: Unit = new Bounds[B] {
    lazy val it = ???  // def or val okay
  }
}
```

Above, the `it` in `$anon` in `x$mcZ$sp` had its owner set to `x$mcZ$sp`
instead of `$anon`. This crashed the typer when it had to retype its
lazy accessor, because there was no `lazy var it` in `$anon`.
Furthermore, the duplicated symbol wasn't being added to the list of
declarations of `$anon`.

Changes:
1) `invalidate` in Duplicators takes an additional parameter which is
the new owner of the new symbol that has to be duplicated. If this parameter
is set to `NoSymbol`, then the new owner is `context.owner`, as before.
2) the newly created lazy val symbol is being added to the list of
declarations of its new owner, provided that the owner is a class.

Review by dragos.
@mention dragos
review by @dragos

Commit:
Add more logging.

Commit:
Fixes SI-4717.

A lazy val declared inside an anonymous class inside a specialized
context no longer crashes Duplicators.

Previously, a duplicated lazy val was assigned to the wrong owner
in Duplicators:

    def x[B >: A]: Unit = new Bounds[B] {
      lazy val it = ???  // def or val okay
    }

Above, the `it` in `$anon` in `x$mcZ$sp` had its owner set to `x$mcZ$sp`
instead of `$anon`. This crashed the typer when it had to retype its
lazy accessor, because there was no `lazy var it` in `$anon$`.
Furthermore, the duplicated symbol wasn't being added to the list of
declarations of `$anon`.

Changes:
1) `invalidate` in Duplicators takes an additional parameter which is
the new owner of the new symbol that has to be duplicated. If this parameter
is set to `NoSymbol`, then the new owner is `context.owner`, as before.
2) the newly created lazy val symbol is being added to the list of
declarations of its new owner.

Removes debugging output from the previous commit.

Review by dragos.
@mention dragos

Commit:
Further fixes SI-4717.

Only adding a lazy val into list of declarations if the owner is a class.

Review by dragos.
@mention dragos

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/118/

Commit:
Fixes previous commit.

Should be checking if the owner of the new symbol is a class, not the new symbol itself.

adriaanm commented:
sorry, you need to submit a new pull request for the build bot to pick it up
once the tests pass and the reviewer gives the green light, I'll merge

axel22 commented:
alright!
